src/api/uploads/FolderUpload.js

```diff
@@ -88,18 +88,69 @@ class FolderUpload {
     }
 
     /**
-     * Upload folderNodes
+     * Build folder tree from dataTransferItems
+     *
+     * @param {Array<DataTransferItem | UploadDataTransferItemWithAPIOptions>} dataTransferItems
+     * @returns {Promise<any>}
+     */
+    async buildFolderTreeFromDataTransferItems(
+        dataTransferItems: Array<DataTransferItem | UploadDataTransferItemWithAPIOptions>
+    ) {
+        dataTransferItems.forEach((itemData) => {
+            const item = getDataTransferItem(itemData);
+            const apiOptions = getDataTransferItemAPIOptions(itemData);
+            const entry = getEntryFromDataTransferItem(item);
+            const { name } = entry;
+
+            this.folders[name] = this.createFolderUploadNode(name, apiOptions, entry);
+        });
+    }
+
+    /**
+     * Create a FolderUploadNode instance
+     *
+     * @param {string} name
+     * @param {Object} apiOptions
+     * @param {FileSystemFileEntry} [entry]
+     * @returns {FolderUploadNode}
+     */
+    createFolderUploadNode(name: string, apiOptions: Object, entry?: FileSystemFileEntry): FolderUploadNode {
+        return new FolderUploadNode(
+            name,
+            this.uploadFile,
+            this.addFolderToQueue,
+            apiOptions,
+            {
+                ...this.baseAPIOptions,
+                ...apiOptions
+            },
+            entry
+        );
+    }
+
+    /**
+     * Upload folders
      *
      * @public
      * @param {Object} Options
      * @param {Function} options.errorCallback
-     * @returns {void}
+     * @returns {Promise<any>}
      */
-    upload({ errorCallback }: { errorCallback: Function }): void {
-        // $FlowFixMe
-        Object.values(this.folderNodes).forEach((node: FolderUploadNode) => {
-            node.upload(this.destinationFolderId, errorCallback, true);
-        });
+    async upload({
+        errorCallback,
+        successCallback
+    }: {
+        errorCallback: Function,
+        successCallback: Function
+    }): Promise<any> {
+        await Promise.all(
+            // $FlowFixMe;
```

Comment:
Whats the flow issue here?

Reply:
`Object.values()` return `Array<mixed>`, going to cast instead